packages/jaeger-ui/src/components/common/CopyIcon.test.js

```diff
@@ -28,43 +30,32 @@ describe('<CopyIcon />', () => {
     tooltipTitle: 'tooltipTitleValue',
   };
   let copySpy;
-  let wrapper;
+  let user;
 
   beforeAll(() => {
     copySpy = jest.spyOn(copy, 'default');
   });
 
   beforeEach(() => {
     copySpy.mockReset();
-    wrapper = shallow(<CopyIcon {...props} />);
+    user = userEvent.setup();
+    render(<CopyIcon {...props} />);
   });
 
   it('renders as expected', () => {
-    expect(wrapper).toMatchSnapshot();
+    expect(screen.getByRole('button'));
   });
 
-  it('updates state and copies when clicked', () => {
-    expect(wrapper.state().hasCopied).toBe(false);
+  it('updates state and copies when clicked', async () => {
+    // Intially there should be nothing in the clipboard, and thus the tooltip should have the default title
+    await user.hover(screen.getByRole('button'));
+    await waitFor(() => expect(screen.getByText('tooltipTitleValue')).toBeInTheDocument());
     expect(copySpy).not.toHaveBeenCalled();
 
-    wrapper.find(Button).simulate('click');
-    expect(wrapper.state().hasCopied).toBe(true);
+    // Simulate the click
+    await user.click(screen.getByRole('button'));
+    await user.hover(screen.getByRole('button'));
+    expect(screen.getByText('Copied')).toBeInTheDocument();
```

Comment:
is this test going to he flaky? Once `user.click` happens, the Copied title only remains visible for a certain time - during CI the execution of this code might be suspended (CPU scheduling etc) and resumed after the timeout when "Copied" is no longer displayed, causing this assertion to fail. Some stateful test would be more reliable here, i.e. validating that a certain function was called (and separately testing that the tooltip changes based on state by _controlling the state_, not the user click behavior)

Reply:
- The test wouldn't be flaky as the tooltip would be visible for time-periods in magnitude of seconds, where as the suspension of the test shouldn't be more than microseconds. We are using `async` functions here as there is some time needed by React to render the elements into the DOM, which is not immediate. This is even true for the functionality in production code, and thus we needn't worry about it.

- I get the sentiment to feel more confident about the component using stateful tests, but the same goes against the philosophy of RTL itself. RTL is so heavily invested on testing user interactions, that it offers no utilities to get access to the `props` or `state` of the component after it's render. All you can do is use `querySelectors` to check for the appearance of elements. styles and classes.

Reply:
1. It's possible for VM running CI to get suspended for seconds. However, on a second glance, the change in text does not happen on the timer, but on leaving the tooltip hover, which doesn't seem like it would happen in this test, so it's probably ok.
2. I think you can always construct code in a way to work around that. For example, the state change here happens via `handleTooltipVisibilityChange` function, that could be mocked if necessary.

Anyway, I think we're good because of (1) above.

Reply:
Actually, you don't seem to test that the text will change back from Copied to Copy, is that something we can do by undoing hover state?

Reply:
> Actually, you don't seem to test that the text will change back from Copied to Copy, is that something we can do by undoing hover state?

Yes. I was thinking of adding that myself when I saw that I missed the `handleTooltipVisibilityChange` function in coverage. Will update that as well.